### PR TITLE
Pt 162578740 fate implementation spike

### PR DESCRIPTION
**Pull request by happi**
Experiment with Fate - transaction engine.
Simple operations, control flow, basic blocks, call/return implemented.

### CODE REVIEW

File: apps/aefate/src/aefa_fate.erl [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+-define(trace(I,S), S).
+-endif.
+
+run(What, Chain) ->
+    EngineState = setup_engine(What, Chain),
+    execute(EngineState).
+
+execute(EngineState) ->
+    Instructions = current_bb(EngineState),
+    loop(Instructions, EngineState).
```
Comment (ThomasArts) [inline apps/aefate/src/aefa_fate.erl]: It seems a bit error prone to have 2 copies of current_bb, one update in the loop in EngineState and one updated in the recurser...

Comment (happi) [reply apps/aefate/src/aefa_fate.erl]: Yes. Not so happy with that. Ill try to fix it.

File: apps/aefate/src/aefa_fate.erl [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+    case maps:get(Name, Functions, void) of
+        void ->  throw({error, {trying_to_call_function, Name}});
+        {Signature, _Code} -> Signature
+    end.
+
+check_signature({ArgTypes, _RetSignature},
+                #{ accumulator := Acc
+                 , accumulator_stack := Stack}
+               =_EngineState) ->
+    case check_arg_types(ArgTypes, [Acc, Stack]) of
```
Comment (UlfNorell) [inline apps/aefate/src/aefa_fate.erl]: `[Acc | Stack]`

File: apps/aefate/src/aefa_fate.erl [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+
+push_arguments(Args, #{ accumulator := X
+                      , accumulator_stack := XS} = ES) ->
+    push_arguments(lists:reverse(Args), X, XS, ES).
+
+push_arguments([], Acc, Stack, ES) ->
+    ES#{ accumulator := Acc
+       , accumulator_stack := Stack};
+push_arguments([A|As], Acc, Stack, ES ) ->
+    push_arguments(As, A, [Acc, Stack], ES).
```
Comment (UlfNorell) [inline apps/aefate/src/aefa_fate.erl]: `[Acc | Stack]` ?

File: apps/aefate/src/aefa_fate.erl (hunk 2 above)
Comment (happi) [reply apps/aefate/src/aefa_fate.erl]: Indeed.

File: apps/aefate/src/aefa_fate.erl [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+loop(Instructions, EngineState) ->
+    case step(Instructions, EngineState) of
+        {stop, FinalState} ->
+            FinalState;
+        {jump, BB, NewState} ->
+            {NewInstructions, State2} = jump(BB, NewState),
+            loop(NewInstructions, State2)
+    end.
+
+step([], EngineState) ->
```
Comment (gorillainduction) [inline apps/aefate/src/aefa_fate.erl]: Do we really want fall-through? I could imagine always having a jump instruction at the end of BBs.

**Comment by lucafavatella** [comment]
I took the liberty of labelling as WIP as I understand not for merging yet.

**Pull request closed without merging**